src/Smuggler/Plugin.hs

```diff
@@ -29,12 +44,67 @@ smugglerPlugin _ modSummary tcEnv = do
     let modulePath = ms_hspp_file modSummary
 
     fileContent <- liftIO $ BS.readFile modulePath
-    liftIO $ void $ hashStore cacheDir (smuggling modulePath) (takeFileName modulePath, fileContent)  -- TODO: remove takeFileName
+    uses <- readMutVar (tcg_used_gres tcEnv)
+    liftIO $ void $ hashStore cacheDir (smuggling uses modulePath) (takeFileName modulePath, fileContent)  -- TODO: remove takeFileName
 
     pure tcEnv
   where
-    smuggling :: FilePath -> ByteString -> IO ByteString
-    smuggling modulePath _ = do
+--    smuggling :: FilePath -> ByteString -> IO ByteString
+    smuggling uses modulePath _ = do
+        -- 1. find positions of unused imports
+        let user_imports = filter (not . ideclImplicit . unLoc) (tcg_rn_imports tcEnv)
+        let usage = findImportUsage user_imports uses
+        let unusedPositions = concatMap unusedLocs usage
+--        debugAST unusedPositions
+
+        -- 2. Remove positions of unused imports from annotations.
         (anns, ast) <- runParser modulePath  -- TODO: don't read file, use given ByteString
-        putStrLn $ exactPrint ast $ removeAnnAtLoc 4 31 anns
+        let purifiedAnnotations = foldl' (\ann (x, y) -> removeAnnAtLoc x y ann) anns unusedPositions
+        putStrLn $ exactPrint ast purifiedAnnotations
+
+        -- 3. Return empty ByteString
         pure ""
+
+unusedLocs ::ImportDeclUsage -> [(Int, Int)]
+unusedLocs (L (RealSrcSpan loc) decl, used, unused)
+    -- Do not warn for `import M ()`
+    | Just (False, L _ []) <- ideclHiding decl
+    = []
+
+    -- Note [Do not warn about Prelude hiding]
+    -- TODO: add ability to support custom prelude
+    | Just (True, L _ hides) <- ideclHiding decl
+    , not (null hides)
+    , pRELUDE_NAME == unLoc (ideclName decl)
+    = []
+
+    -- Nothing used; drop entire decl
+    -- TODO: drop every line of multiline import
+    -- TODO: optimize
+    | null used = map (srcSpanStartLine loc,) [1..100]
```

Comment:
Need to investigate more how to handle this, because it can be a multiple-line import. In this case, we first need to check the end location of the whole import (I guess it's possible to do with only checking the element with starting position on column 1 of this line).

Reply:
@vrom911 Yeah, we need to investigate this. I propose to do it under separate issue. I guess we can check start and end line of import and just call this with every line for now.